### PR TITLE
Gradient loss for mvsnet

### PR DESCRIPTION
**Pull request by cpheinrich**
This PR adds a gradient loss term for MVSNet. As a bonus I also threw in a gaussian loss term. The different loss parameters can be tuned using flags in the train.py file.

### CODE REVIEW

File: mvsnet/train.py [hunk 1]
```diff
@@ -474,30 +494,38 @@ def train():
...
 
                     # print info
                     if step % FLAGS.display == 0:
                         print(Notify.INFO,
                               'epoch, %d, step %d, total_step %d, loss = %.4f, (< 1px) = %.4f, (< 3px) = %.4f (%.3f sec/step)' %
                               (epoch, step, total_step, out_loss, out_less_one, out_less_three, duration), Notify.ENDC)
+                    if out_loss == np.nan:
+                        # Stop the job if we get a nan in the loss function
+                        sys.exit(1)
+
```
Comment (miaoshasha) [inline mvsnet/train.py:508]: :+1:

File: mvsnet/loss.py [hunk 2]
```diff
@@ -12,48 +12,151 @@
...
         shape = tf.shape(y_pred)
         interval = tf.reshape(interval, [shape[0]])
+        # mask_true is a tensor of 0's and 1's, where 1 is for valid pixels and 0 for invalid pixels
         mask_true = tf.cast(tf.not_equal(y_true, 0.0), dtype='float32')
-        denominator = tf.abs(tf.reduce_sum(mask_true, axis=[1, 2, 3])) + 1e-6
-        if beta != 1.0:
-            denominator = tf.math.pow(denominator, beta)
-        # Below we assume the random error in y_true increases linearly with distance
-        # and that the error in y_true is 5mm at a distance of 1 meter
+        # The number of valid pixels in the depth map -- used for taking the average over only valid pixels
```
Comment (miaoshasha) [inline mvsnet/loss.py:57]: not sure about the specific dataset you're using, but is there a value that's set when the depth value is too high, e.g. for the sky? do we want to mask out those pixels as well?

File: mvsnet/loss.py [hunk 3]
```diff
@@ -12,48 +12,151 @@
...
+        num_valid_pixels = tf.abs(tf.reduce_sum(
+            mask_true, axis=[1, 2, 3])) + 1e-6
```
Comment (miaoshasha) [inline mvsnet/loss.py:59]: nit pick: tf.abs is probably not needed here.

File: mvsnet/loss.py [hunk 4]
```diff
@@ -12,48 +12,151 @@
...
+def gaussian_loss(y_true, y_pred, interval, eta):
+    """ non zero mean absolute loss for one batch
+
+    This function parameterizes a loss of the form
+
+    Loss = - exp(- x ^ 2 / 2*sigma ^ 2)
+
+    where x = y_true - y_pred and
+    sigma = eta * y_true
+
```
Comment (miaoshasha) [inline mvsnet/loss.py:102]: is this based on the assumption that the noise is proportional to its value?

File: mvsnet/loss.py [hunk 5]
```diff
@@ -12,48 +12,151 @@
...
+    """
+
+    with tf.name_scope('MAE'):
+        shape = tf.shape(y_pred)
+        interval = tf.reshape(interval, [shape[0]])
+        # mask_true is a tensor of 0's and 1's, where 1 is for valid pixels and 0 for invalid pixels
+        mask_true = tf.cast(tf.not_equal(y_true, 0.0), dtype='float32')
+        # The number of valid pixels in the depth map -- used for taking the average over only valid pixels
+        num_valid_pixels = tf.abs(tf.reduce_sum(
+            mask_true, axis=[1, 2, 3])) + 1e-6
```
Comment (miaoshasha) [inline mvsnet/loss.py:119]: same here, tf.abs

**Review by miaoshasha — APPROVED** [review]
lgtm

File: mvsnet/loss.py (hunk 4 above)
Comment (cpheinrich) [reply mvsnet/loss.py:102]: essentially, ya, its based on the notion that estimating depth at further distances is harder, and by normalizing by distance you can compare an error at 1m with an error at 10m

File: mvsnet/loss.py (hunk 2 above)
Comment (cpheinrich) [reply mvsnet/loss.py:57]: invalid depths (could be because its too high or whatever), are all set to a value of 0

File: mvsnet/loss.py (hunk 3 above)
Comment (cpheinrich) [reply mvsnet/loss.py:59]: ya good point
